src/grafter/matcha/spec.clj

```diff
@@ -0,0 +1,69 @@
+(ns grafter.matcha.spec
+  (:require [clojure.spec.alpha :as s]
+            [clojure.spec.gen.alpha :as g]))
+
+(defmacro ^:private when-available [syms & body]
+  (when (every? some? (map resolve syms))
+    `(do ~@body)))
+
+(try
+  ;; avoid issue: https://github.com/Swirrl/matcha/issues/5
+  (require '[grafter.rdf.protocols :as gp])
+  (import '[grafter.rdf.protocols RDFLiteral LangString Quad])
+
+  (catch java.io.FileNotFoundException _))
+
+(s/def ::uri uri?)
+
+(s/def ::bnode keyword?)
+
+(s/def ::subject (s/or :uri ::uri :bnode ::bnode))
+
+(s/def ::predicate (s/or :uri ::uri :bnode ::bnode))
```

Comment:
In RDF bnodes are only valid identifiers in object or subject position, not predicate.

The confusion here is probably due to these specs being written, perhaps more with an eye towards grafter itself than matcha.  I wrote them initially as an experiment/exercise on a train journey, and mainly pushed them to talk about a few ideas with Lee.

Anyway to explain the `::bnode` thing.  Matcha itself doesn't directly require grafter.  The motivation for this was partly because I thought it might be nice to drop the JVM dependency and target Matcha towards cljs too.  Not that we have a direct need for this right now, but I thought it might give us some more flexibility on the frontend.  

Regarding grafter, it essentially provides bidirectional type conversions from sesame/rdf4j types to more native clojure values and vice versa.  Sesame bnode objects are then represented in clojure and grafter Quads as clojure keywords.  I made this decision many years ago, and kind of regret it a little now.  It's probably quite straightforward to replace these at some point with a special bnode `deftype`, and free keywords for other things.  Particularly since spec now adds extra capabilities to keywords, and namespaced keywords are much more prevalent.

For example one tension we currently have, is when using `matcha/construct` whether to construct objects out with keywords or URIs for example is it better to do this:

```clojure
(construct {:label ?label, :uri uri, :description ?description} 
                [[uri a qb:DataSet] 
                 [uri rdfs:label ?label] 
                 [uri dcat:description ?description]])
```

or this:

```clojure
 (require '[grafter.rdf.vocabularies.rdf :refer [rdfs:label])
 (require '[grafter.rdf.vocabularies.dcat :refer [dcat:description])
 
(construct {:grafter.rdf/uri uri,
                    rdfs:label ?label
                    dcat:description ?description } 
                [[uri a qb:DataSet] 
                 [uri rdfs:label ?label] 
                 [uri dcat:description ?description]])
```

or more succienctly overselect slightly the whole resource with:

```clojure
(construct {:grafter.rdf/uri uri,
                    ?p ?o } 
                [[uri ?p ?o]])
```

The benefit of the first style is that your clojure object uses keywords, which are easily spec'd with `s/keys` etc.  The disadvantage is that you've introduced another level of mapping indirection which historically hasn't been found to be that useful, and has been a bit of a source of bugs e.g. one person's `:label` is another persons `:name` etc.  

The alternative approaches both use the URI objects themselves, which means the vocab terms flow from the queries to the view.  i.e. you can construct the terms out with the vocab URIs you want to use close to the database, and flow them through to the front end with a minimal amount of remapping.  (You can choose to remap but still keep using URIs too - for example you can do this with a SPARQL (or matcha) CONSTRUCT:

```sparql
PREFIX foo: <http://publishmydata.com/def/foo#>
PREFIX : <http://publishmydata.com/def/bar#>

CONSTRUCT { :my-resource foo:name ?label } 
WHERE { 
  :my-resource rdfs:label ?label .
}
```

Which effectively renames the label predicate. 

The disadvantage of this is that URI's are uglier than keywords, and create more noise when pprinting at a REPL, and that `s/keys` only works with clojure keywords.  I asked Alex Miller, if it was possible to extend keywordiness to other types, and he said it's possible but essentially involves writing completely new implementations so we're talking a new `s/key-uris` and possibly `s/def-uri`.  Incidentally it's worth mentioning that you can still destructure URI maps directly with code like this:

```clojure
(def rdfs:label (URI. "http://.../label"))

(let [{label rdfs:label} {rdfs:label "hello"}]
   label) ;; => "hello"
```

For simple cases it's ok doing this remapping; but I feel like whenever we do it we just end up rebuilding the same kind of machinery but slightly differently each time; and that it's better to just let the grafter rdf types just flow straight to the UI.  As that way we can more consistently handle things like language strings and priorities on labels etc.  i.e. I think there are benefits in using RDF end-end.  

So one area I've been keen to explore is tying to bridge this divide, by either making keywords behave more like URIs, or by making URI's behave more like keywords.  For example one thing we could at some point explore is using keywords to represent prefixed RDF vocabularies, e.g. we could provide some low level functions like:

```clojure
(def prefix-map {:rdfs (URI. "http://www.w3.org/2000/01/rdf-schema#")})

(keyword->uri :rdfs/label) ;; => (URI. "http://www.w3.org/2000/01/rdf-schema#label")
```

And then provide native support for these in standard utilities like matcha.  I think these tools would need to be opt-in/outable as they complicate a few things like:

```clojure
(= :rdfs/label (URI. "http://www.w3.org/2000/01/rdf-schema#label")) ;; => false
```

But it would let us represent resource objects like:

```clojure
{:rdfs/label "my label"
 :grafter.rdf/uri (URI. "http://foobar/baz")
}
```

Typically I'd expect the keywords would just be used for vocab terms, i.e. RDF predicates and classes, rather than user facing resources.

Anyway these kinds of changes are probably a while off, but it's worth keeping them in mind, as I suspect things like this are fertile ground for making development much smoother.

Reply:
Maybe a data-reader like `#uri rdfs/label`or even `#rdfs label` rather than this keywordyness? (dunno how many of these prefix maps there are so that might get a bit unwieldy).

If it's just a case of shorthand for URI building.

Reply:
Yes data readers are definitely the other option, and something we've been talking about too... just forgot to mention it...  Though actually I hadn't considered literally managing the prefix maps as the data_readers.edn files themselves.  That might well be the best option.

One issue is that the set of prefixes may potentially change per customer etc...  Though we do have a [centralised vocabulary project/dependency](https://github.com/Swirrl/grafter-vocabularies/) that would be a natural place to share something like this.

We'd also want to extend the print-methods, we've done this in a few projects, but it'd be good to centralise it.

We could perhaps provide some tooling to generate a `data_readers.edn` file from some config in the builds etc too...  There are quite a lot of mappings in aggregate; probably dozens of vocab prefixes which in turn mean hundreds of CURI mappings.

I can't recall the semantics about how `data_readers.edn` files are discovered on the classpath and merged in the case of conflicts from libraries etc... but something like this could certainly be made to work.

Reply:
I think they are merged, hash-map style, in the order of appearance on the classpath, so not exactly elegant. As far as I remember, you can manage it directly at runtime by rebinding the `clojure.core/*data-readers*` root, but also not that nice an option.

I guess the "namespaced reader tags only" "restriction" in user code (that isn't actually enforced) was probably there to prevent clashes.

Reply:
> I guess the "namespaced reader tags only" "restriction" in user code (that isn't actually enforced) was probably there to prevent clashes.

Yeah it's just a convention to encourage namespacing and composition of edn extensions.  I've filed issues with projects in the past (aero and integrant) as I wanted to use them together and they both defined `#ref` as a tag, which you can work around but it's better to use namespaced tags.  That's one of the concerns with using them as you suggest, and then the following sort of thing starts to look a bit more noisy, but c'est la vie!

`#swirrl/uri rdfs/label`

or perhaps

`#vocab/rdfs label`
`#vocab/qb Dataset`
...